pulsar/permission_grant.go

```diff
@@ -0,0 +1,78 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package pulsar
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/hashcode"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/streamnative/pulsarctl/pkg/pulsar/common"
+	"github.com/streamnative/terraform-provider-pulsar/types"
+)
+
+func permissionGrantToHash(v interface{}) int {
+	var buf bytes.Buffer
+	m := v.(map[string]interface{})
+
+	buf.WriteString(fmt.Sprintf("%s-", m["role"].(string)))
+	buf.WriteString(fmt.Sprintf("%s-", m["actions"].([]string)))
+
+	return hashcode.String(buf.String())
+}
+
+func unmarshalPermissionGrants(v []interface{}) ([]*types.PermissionGrant, error) {
+	permissionGrants := make([]*types.PermissionGrant, 0, len(v))
+	for _, grant := range v {
+		data := grant.(map[string]interface{})
```

Comment:
have you looked into mapstructure.decode

https://godoc.org/github.com/mitchellh/mapstructure#Decode

Reply:
Ah nice! That said, I do have some reservations around using that in this PR - particularly:

1. This code here was pretty much just migrated as-is over here from where it previously was in `resource_pulsar_namespace.go`
2. I'm hesitant to introduce a new direct dependency into the project

Generally, with both of the above - I think it boils down to a preference of leaning towards not introducing too many changes all at once into someone else's project

Reply:
That makes sense.